UPD: update to the NonlinearExpr syntax of JuMP v1.15

odow opened the pull request:
Closes #453

The tests haven't finished running locally, so there might be some more changes needed, but this is a start. I'll comment in-line about some formatting changes we could do here, or in a separate PR.

odow commented on src/core/base.jl:
```diff
@@ -69,22 +69,6 @@ function _check_var_keys(vars, keys, var_name, comp_name)
-            JuMP.@constraint($model, $expr)
-        end
-    end)
-end
```
This PR is breaking because it removes this macro which was documented.

odow commented on src/form/en_shared.jl:
```diff
@@ -133,7 +133,7 @@ function constraint_mc_current_balance(pm::RectangularVoltageExplicitNeutralMode
 
     for (idx, t) in ungrounded_terminals
-        @smart_constraint(pm.model,  [cr, crd, crg, crs, crsw, crt, vr],
+        JuMP.@constraint(pm.model,
```
As currently formatted, the body of the constraint is indented by a lot. (It used to match the opening `[` of the second argument. I can either leave to minimize the diff of this PR, or fix. Not sure which is best.

odow commented:
@ccoffrin (or anyone): are the failing tests global solutions (and there's a bug somewhere)? Or are these local minima, and by changing the order of constraints etc we now find new solutions?

ccoffrin commented:
Good question. I am not that familiar with how stable the tests are in terms of local minima. The failing tests do appear to be consolidated in the non-convex formulations, so it is a reasonable hypothesis.  We will need one of the primary developers to look into it to assess.

odow commented:
Hey @pseudocubic any chance you could take a look at this? Or point me in the right direction?

pseudocubic commented:
@odow we have been looking at this but still haven't identified an obvious pattern in the failing tests, and based on the changes in PMD and JuMP 1.15 don't understand what would have caused all the failures we're seeing.

odow commented on Project.toml:
```diff
@@ -23,22 +23,24 @@ Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
+Ipopt = "1"
 JSON = "0.18, 0.19, 0.20, 0.21"
-JuMP = "0.22, 0.23, 1"
+JuMP = "1.15"
```
```suggestion
JuMP = "1.23.2"
```

odow replied:
Waiting for https://github.com/JuliaRegistries/General/pull/115113

odow commented:
Now we're getting somewhere. These two failures just look flakey? One looks like a tolerance issue, and the other is Ipopt not succeeding.

odow commented on test/transformer.jl:
```diff
@@ -369,7 +369,7 @@
         @testset "3w_dyy_1" begin
             result = solve_mc_opf(ut_trans_3w_dyy_1, SOCConicUBFPowerModel, scs_solver; solution_processors=[sol_data_model!], make_si=false)
-            @test norm(result["solution"]["bus"]["3"]["vm"]-[0.93180, 0.88827, 0.88581], Inf) <= 7.2E-2
+            @test norm(result["solution"]["bus"]["3"]["vm"]-[0.93180, 0.88827, 0.88581], Inf) <= 1E-3
```
```suggestion
            @test norm(result["solution"]["bus"]["3"]["vm"]-[0.93180, 0.88827, 0.88581], Inf) <= 1E-1
```

odow commented on test/transformer.jl:
```diff
@@ -369,7 +369,7 @@
         @testset "3w_dyy_1" begin
             result = solve_mc_opf(ut_trans_3w_dyy_1, SOCConicUBFPowerModel, scs_solver; solution_processors=[sol_data_model!], make_si=false)
-            @test norm(result["solution"]["bus"]["3"]["vm"]-[0.93180, 0.88827, 0.88581], Inf) <= 7.2E-2
+            @test norm(result["solution"]["bus"]["3"]["vm"]-[0.93180, 0.88827, 0.88581], Inf) <= 1E-1
```
@pseudocubic I relaxed this tolerance.  The other tests above this also use a mix of tolerances, so it seems a bit flakey.

odow commented on test/en_opf_bounds.jl:
```diff
@@ -76,7 +76,9 @@ end
-        sol_pmd = calc_sol_pmd(data_math, ACRENPowerModel; optimizer=optimizer_with_attributes(Ipopt.Optimizer, "mu_strategy"=>"adaptive", "sb"=>"yes", "print_level"=>0))
+        # This test is flakey.
+        mu_strategy = Sys.islinux() ? "adaptive" : "monotone"
+        sol_pmd = calc_sol_pmd(data_math, ACRENPowerModel; optimizer=optimizer_with_attributes(Ipopt.Optimizer, "mu_strategy"=>mu_strategy, "sb"=>"yes", "print_level"=>0))
```
@pseudocubic I changed the `mu_strategy` to the default except for linux because this test seemed to fail sporadically. I assume that the non-default option was added precisely because it's required sometimes (but not all the time).

ccoffrin commented:
Given all tests are passing, is there any reason not to merge and tag?

odow commented:
From my side, I'm good.

pseudocubic commented on Project.toml:
```diff
@@ -23,22 +23,24 @@ Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
 CSV = "0.8.5, 0.9, 0.10"
+Dates = "1.6"
 FilePaths = "0.8.3"
 Glob = "1.3"
 Graphs = "1"
 InfrastructureModels = "0.7.3, 0.7.5"
-Ipopt = "0.9, 1.0.2, 1.1"
+Ipopt = "1"
 JSON = "0.18, 0.19, 0.20, 0.21"
-JuMP = "0.22, 0.23, 1"
+JuMP = "1.23.2"
 LoggingExtras = "0.4.7, 1"
 PolyhedralRelaxations = "0.3.5"
 SCS = "0.9, 1.0, 1.1"
+SparseArrays = "1.6"
 SpecialFunctions = "2"
+Test = "1.6"
```
Is there a reason Dates, SparseArrays and Test were added to compat? My impression was that these don't need to be explicitly listed because of the julia compat entry.

pseudocubic reviewed (changes requested):
I have re-reviewed the PR from scratch, and left a comment on some lines added to Project.toml. 

Also, please add an update to the CHANGELOG

Otherwise, it looks good, thanks for helping us to track down the issue in JuMP and make these changes to PMD!

odow commented:
I've added the CHANGELOG entry.

For the compat bounds: future versions of Julia are going to decouple stdlib versions from the Julia version, so best practice is to add compat bounds. But there was a bug in Julia 1.6 that prevented them from enforcing this rule all packages at registration time.

I've removed for now, but in some future version, we'll likely need to add them back. (Probably when/if you set `julia = "1.10"`.)

pseudocubic commented:
Looks good! Thanks for the info about stdlib versions getting decoupled, I didn't know that, so I'll keep an eye out in the future.

Thanks again for the contribution!
